src/mame/altos/altos586_hdc.cpp

```diff
@@ -0,0 +1,380 @@
+// license:BSD-2-Clause
+// copyright-holders:Lubomir Rintel
+
+/***************************************************************************
+
+    Altos 586 Hard Disk Controller emulation
+
+    The controller is based around the 8089 I/O Processor.
+    Its memory bus is connected (via modified multibus) to the main memory,
+    shared with the main 8086 processor. The I/O bus contains the internal
+    controller registers and a 16K SRAM for caching the I/O programs
+    I/O parameter blocks (command blocks) and staging the sector data.
+
+    The formatting is custom, enforces 16 sectors per tracks.
+    Heads/Cylinder combinations used by different Altos 586 models
+    as supported by the diagnostic software (and XENIX, I think):
+
+    Model        Heads   Cylinders
+    ACS586-10    4       306
+    ACS586-20    6       306
+    ACS586-30    6       512
+    ACS586-40    8       512
+    H-H 20 MB    4       612
+
+    Literature:
+
+    [1] Notes on the Altos 586 Computer & Firmware disassembly
+        https://github.com/lkundrak/altos586/
+
+    [2] Preliminary 8600 User Manual
+        Section 7. System Specs.
+	4.5. Rigid Disk Controllers and Interface
+	Specification Revision 4.2, May 27, 1982, Page 52-62
+
+    [3] Preliminary 8600 User Manual
+        Section 7. System Specs.
+	A. Rigid Disk Controller Documentation
+	Specification Revision 4.2, May 27, 1982, Page 95-108
+
+    Note: The Altos 8600 manuals [2] and [3] describe a controller
+    wired differently (the 8089 seems to be in local mode, not remote),
+    the register addresses are different, but their function and command
+    set seems very different. I've written this before discovering those
+    manuals, so the command names are slightly different. My intent is to
+    eventually align that and review some of my assumptions too.
+
+    TODO: How is the command completion interrupt signalled?
+    I've only verified it to work with the diags floppy and that
+    one is not interrupt-driven. XENIX almost certainly is.
+
+***************************************************************************/
+
+#include "emu.h"
+#include "altos586_hdc.h"
+
+#define VERBOSE 0
+
+DEFINE_DEVICE_TYPE(ALTOS586_HDC, altos586_hdc_device, "altos586_hdc", "Disk Controller board for Altos 586")
+
+altos586_hdc_device::altos586_hdc_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, ALTOS586_HDC, tag, owner, clock)
+	, m_bus(*this, finder_base::DUMMY_TAG)
+	, m_iop(*this, "iop")
+	, m_hdd0(*this, "drive0")
+	, m_hdd1(*this, "drive1")
+{
+}
+
+bool altos586_hdc_device::sector_exists(uint8_t index)
+{
+	if (!m_geom[m_drive]) {
+		logerror("drive not present: %d", m_drive);
+	} else if (m_geom[m_drive]->sectorbytes != 512) {
+		logerror("expected 512 bytes per sector, got %d", m_geom[m_drive]->sectorbytes);
```

Comment:
Should this be checked when mounting a disk image rather than on every sector access?